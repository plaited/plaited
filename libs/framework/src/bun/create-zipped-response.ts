export const jsMimeTypes = ['text/javascript;charset=utf-8', 'application/json;charset=utf-8'] as const

export const isJsMimeType = (type: string): type is (typeof jsMimeTypes)[number] =>
  jsMimeTypes.includes(type as (typeof jsMimeTypes)[number])

export const createZippedResponse = (str: string, type: (typeof jsMimeTypes)[number]) => {
  const compressed = Bun.gzipSync(Buffer.from(str))
  return new Response(compressed, {
    headers: {
      'content-type': type,
      'content-encoding': 'gzip',
    },
  })
}
