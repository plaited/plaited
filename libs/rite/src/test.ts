/// <reference lib="dom" />
/// <reference lib="dom.iterable" />

import { TestResultError, TestResult, BrowserSessionResult } from '@web/test-runner-core/browser/session.js'
import { assert, Assertion, AssertionError } from './assert.js'

type TestCallback = (t: Assertion) => Promise<void> | void

interface Test {
  (name: string, cb: TestCallback): void
  skip: (name: string, cb: TestCallback) => void
}

type TestCase = () => Promise<boolean> | boolean

class TestRunner {
  #completedTests = 0
  #failedTests = 0
  #errors: TestResultError[] = []
  #cases: TestCase[] = []
  #timeout = 5_000
  #tests: TestResult[] = []
  updateTimeout(time: number) {
    this.#timeout = time
  }
  get timeout() {
    return this.#timeout
  }
  addCase(test: TestCase) {
    this.#cases.push(test)
  }
  addResult(result: TestResult) {
    this.#tests.push(result)
  }
  addError(obj: TestResultError) {
    this.#errors.push(obj)
  }
  async run(name: string) {
    for (const testCase of this.#cases) {
      try {
        const passed = await testCase()
        this.#completedTests++
        if (!passed) {
          this.#failedTests++
        }
      } catch (error) {
        console.error(error)
      }
    }

    if (this.#completedTests === this.#cases.length) {
      const results: BrowserSessionResult = {
        passed: !this.#failedTests,
        testResults: {
          name,
          suites: [],
          tests: this.#tests,
        },
        errors: this.#errors,
      }
      return results
    } else {
      throw new AssertionError(`Not all test completed`)
    }
  }
}

declare global {
  interface Window {
    __rite_test_runner?: TestRunner
  }
}

window.__rite_test_runner = new TestRunner()

const throwTimeoutError = async () => {
  const timeout = window.__rite_test_runner.timeout
  await assert.wait(timeout)
  throw new Error(`test takes longer than ${timeout}ms to complete`)
}

export const test: Test = (name, callback) => {
  window.__rite_test_runner.addCase(async () => {
    const startTime = performance.now()
    const result: TestResult = {
      name,
      passed: true,
      skipped: false,
    }
    try {
      await Promise.race([callback(assert), throwTimeoutError()])
      const msg = `✓ ${name}`
      // eslint-disable-next-line no-console
      console.log(msg)
    } catch (error) {
      const msg = `✗ ${name}`
      // eslint-disable-next-line no-console
      console.log(msg)
      result.passed = false
      if (error instanceof AssertionError) {
        const obj: Pick<TestResultError, 'actual' | 'expected' | 'message'> = JSON.parse(error.message)
        result.error = { name, stack: `${error.stack}`, ...obj }
        window.__rite_test_runner.addError({ name, ...obj })
      } else {
        throw new Error(error)
      }
    }
    result.duration = performance.now() - startTime
    window.__rite_test_runner.addResult(result)
    return result.passed
  })
}
test.skip = (name: string, _: TestCallback) => {
  window.__rite_test_runner.addCase(() => {
    const result = {
      name,
      passed: true,
      skipped: true,
    }
    window.__rite_test_runner.addResult(result)
    return true
  })
}
