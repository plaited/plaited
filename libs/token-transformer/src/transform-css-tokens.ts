import { TransformerParams } from './types.js'
import { defaultCSSFormatters } from './css-tokens/index.js'
import { formatList } from './format-list.js'
import { defaultBaseFontSize } from './constants.js'
import { reduceWhitespace } from '@plaited/utils'

const deduplicate = (css: string) => {
  const regex = /((?:.*:host|:host\([^)]*\))[^{\n]*)\{(\s*[\s\S]*?\s*)\}/gm
  const map = new Map<string, Set<string>>()
  let match: RegExpExecArray | null
  while ((match = regex.exec(css)) !== null) {
    const selector = match[1]
    const rule = reduceWhitespace(match[2])
    const set = map.get(selector)
    if (set) {
      set.add(rule)
      continue
    }
    map.set(selector, new Set<string>([rule]))
  }
  return [...map]
    .flatMap(([key, val]) => {
      return [key, '{', ...val, key.startsWith('@') ? '}}' : '}']
    })
    .join('')
}

/**
 * Transforms design tokens into CSS rules with deduplicated selectors on :host.
 * These rules are to be used with design token custom element and applied to
 * it's shadow root's constructable stylesheet.
 * @param {TransformerParams} params - The parameters for the transformation.
 * @returns {string} The transformed CSS rules with deduplicated selectors.
 */
export const transformCssTokens = ({
  tokens,
  baseFontSize = defaultBaseFontSize,
  formatters = defaultCSSFormatters,
  mediaQueries,
  colorSchemes,
  containerQueries,
}: TransformerParams) => {
  const content = formatList({
    tokens,
    allTokens: tokens,
    baseFontSize,
    formatters,
    mediaQueries,
    colorSchemes,
    containerQueries,
  })
  return deduplicate(content)
}
