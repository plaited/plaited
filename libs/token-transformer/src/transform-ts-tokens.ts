import { mkdir } from 'node:fs/promises'
import { formatList } from './format-list.js'
import { DesignTokenGroup, GetFormatters } from '@plaited/token-types'

export const transformTsTokens = async ({
  tokens,
  output,
  baseFontSize,
  formatters,
}: {
  tokens: DesignTokenGroup;
  output: string;
  baseFontSize: number;
  formatters: GetFormatters;
}) => {
  await mkdir(output, { recursive: true })
  const content = formatList({
    tokens,
    allTokens: tokens,
    baseFontSize,
    formatters,
  })
  await Bun.write(`${output}/tokens.ts`, content)
}
