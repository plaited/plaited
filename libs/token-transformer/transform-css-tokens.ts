import { formatList } from './format-list.ts'
import { DesignTokenGroup, GetFormatters } from '../token-types.ts'
import { combineDuplicatedSelectors, postcss } from '../deps.ts'

export const transformCssTokens = async ({
  tokens,
  outputDirectory,
  baseFontSize,
  formatters,
}: {
  tokens: DesignTokenGroup
  outputDirectory: string
  baseFontSize: number
  formatters: GetFormatters
}) => {
  await Deno.mkdir(outputDirectory, { recursive: true })
  const content = formatList({
    tokens,
    allTokens: tokens,
    baseFontSize,
    formatters,
  })
  const { css } = await postcss([
    combineDuplicatedSelectors,
  ]).process(content, { from: undefined, to: '' })
  await Deno.writeTextFile(`${outputDirectory}/tokens.css`, css)
}
