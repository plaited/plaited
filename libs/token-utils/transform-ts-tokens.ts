import { formatList } from './format-list.ts'
import { DesignTokenGroup, GetFormatters } from './types.ts'

export const transformTsTokens = async ({
  tokens,
  output,
  baseFontSize,
  formatters,
}: {
  tokens: DesignTokenGroup
  output: string
  baseFontSize: number
  formatters: GetFormatters
}) => {
  await Deno.mkdir(output, { recursive: true })
  const content = formatList({
    tokens,
    allTokens: tokens,
    baseFontSize,
    formatters,
  })
  await Deno.writeTextFile(`${output}/tokens.ts`, content)
}
