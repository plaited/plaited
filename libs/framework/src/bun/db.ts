import { Database } from "bun:sqlite";

export const db = new Database(":memory:");
db.query(`create table templates;`);