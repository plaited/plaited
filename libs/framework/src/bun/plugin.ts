import type {  BunPlugin } from "bun";
import { db } from "./db.js";

export const plugin: BunPlugin = {
  name: "Plaited loader",
  setup(build) {
    build.onResolve({ filter: /\/route.tsx?$/ }, async ({ path }) => {
      // mark external all non-template files
    })
    build.onLoad({ filter: /\/route.tsx?$/ }, async ({ path }) => {
      // modify the contents of the file to be an export object.
      return {
        contents,
        loader: "tsx",
      };
    });
  },
}