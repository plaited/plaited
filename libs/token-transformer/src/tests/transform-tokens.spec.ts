import { test, expect } from 'bun:test'
import beautify from 'beautify'
import { transformTsTokens } from '../transform-ts-tokens.js'
import { transformCssTokens } from '../transform-css-tokens.js'
import { tokens } from './sample-tokens.js'
import { defaultTSFormatters, defaultCSSFormatters } from '../index.js'

const mediaQueries = {
  desktop: 'screen and (min-width: 1024px)',
  mobile: 'screen and (max-width: 767px)',
  tv: 'screen and (min-width: 1920px)',
}

test('transform-ts-tokens', () => {
  const content = transformTsTokens({
    tokens,
    baseFontSize: 20,
    formatters: defaultTSFormatters,
    mediaQueries,
  })
  expect(content).toMatchSnapshot()
})

test('transform-css-tokens', () => {
  const content = transformCssTokens({
    tokens,
    baseFontSize: 20,
    formatters: defaultCSSFormatters,
    mediaQueries,
  })
  expect(beautify(content, { format: 'css' })).toMatchSnapshot()
})
