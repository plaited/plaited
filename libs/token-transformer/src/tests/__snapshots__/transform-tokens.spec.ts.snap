// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`transform-ts-tokens 1`] = `
"export const sizeX1 = 'var(--size-x1)'
export const sizeX2 = 'var(--size-x2)'
export const sizeX3 = 'var(--size-x3)'
export const sizeS3 = 'var(--size-s3)'
export const backgroundColorPurpleX1 = 'var(--background-color-purple-x1)'
export const backgroundColorPurpleX2 = 'var(--background-color-purple-x2)'
export const backgroundColorPurpleX3 = 'var(--background-color-purple-x3)'
export const borderRadiusX1 = 'var(--border-radius-x1)'
export const borderRadiusX2 = 'var(--border-radius-x2)'
export const borderRadiusX3 = 'var(--border-radius-x3)'
export const fontSizeX1 = 'var(--font-size-x1)'
export const fontFamilySansSerif = 'var(--font-family-sans-serif)'
export const fontWeightX1 = 'var(--font-weight-x1)'
export const letterSpacingX1 = 'var(--letter-spacing-x1)'
export const lineHeightX1 = 'var(--line-height-x1)'
export const typographyX1 = {
  fontFamily: fontFamilySansSerif,
  fontSize: fontSizeX1,
  fontWeight: fontWeightX1,
  letterSpacing: letterSpacingX1,
  lineHeight: lineHeightX1,
}
export const actionPrimaryRestBorderRadius = borderRadiusX1
export const actionPrimaryRestBackgroundColor = backgroundColorPurpleX2
export const actionPrimaryRestTypography = typographyX1
"
`;

exports[`transform-css-tokens 1`] = `
"@media screen and (min-width: 1920px) {
    :host {
        --size-x1: 0.2rem;
        --size-x2: 0.4rem;
        --size-x3: 0.7rem;
    }
}

:host([data-media-query="tv"]) {
    --size-x1: 0.2rem;
    --size-x2: 0.4rem;
    --size-x3: 0.7rem;
}

@media screen and (max-width: 767px) {
    :host {
        --size-x1: 0.1rem;
        --size-x2: 0.4rem;
        --size-x3: 0.8rem;
    }
}

:host([data-media-query="mobile"]) {
    --size-x1: 0.1rem;
    --size-x2: 0.4rem;
    --size-x3: 0.8rem;
}

:host {
    --size-x1: 0.1rem;
    --size-x2: 0.3rem;
    --size-x3: 0.6rem;
    --size-s3: 0.6rem;
    --background-color-purple-x1: oklch(39% 0.22184604903824146 289.1379241925685 / none);
    --background-color-purple-x2: oklch(45.38% 0.236 304.59 / none);
    --background-color-purple-x3: oklch(48.67% 0.245 310.63 / none);
    --border-radius-x1: 0.15rem;
    --border-radius-x2: 0.25rem;
    --border-radius-x3: 0.35rem;
    --font-size-x1: 0.7rem;
    --font-family-sans-serif: Roboto,sans-serif;
    --font-weight-x1: 700;
    --letter-spacing-x1: normal;
    --line-height-x1: 1;
}

@media screen and (min-width: 1024px) {
    :host {
        --size-x1: 0.1rem;
        --size-x2: 0.3rem;
        --size-x3: 0.6rem;
    }
}

:host([data-media-query="desktop"]) {
    --size-x1: 0.1rem;
    --size-x2: 0.3rem;
    --size-x3: 0.6rem;
}"
`;
